<launch>

  <arg name="port_name" default="/dev/ttyUSB0" />
  <arg name="baud_rate" default="57600" />
  <arg name="namespace" default="sample_robot" />
  <arg name="robot_description_file" default="$(find dynamixel_general_hw)/urdf/sample1.urdf" />
  <arg name="dynamixel_info_file" default="$(find dynamixel_general_hw)/config/sample1/dynamixel_info.yaml" />
  <arg name="controllers_file" default="$(find dynamixel_general_hw)/config/sample1/default_controllers.yaml" />
  <arg name="controllers_to_start" default="joint_state_controller position_joint_trajectory_controller" />

  <arg name="set_robot_description" default="true" />
  <arg name="robot_description_param" default="robot_description" />
  <arg name="launch_robot_state_publisher" default="true" />
  <arg name="required" default="true" />
  <arg name="joint_states_topic" default="joint_states" />
  <arg name="control_rate" default="20" />
  <arg name="calculate_effort" default="true" />
  <arg name="launch_rviz" default="false" />
  <arg name="rvizconfig" default="$(find dynamixel_general_hw)/config/sample_robot.rviz" />

  <group ns="$(arg namespace)">

    <param if="$(arg set_robot_description)"
           name="$(arg robot_description_param)"
           command="$(find xacro)/xacro $(arg robot_description_file)" />

    <node if="$(arg launch_robot_state_publisher)"
          name="robot_state_publisher"
          pkg="robot_state_publisher" type="robot_state_publisher">
      <remap from="robot_description" to="$(arg robot_description_param)" />
      <remap from="joint_states" to="$(arg joint_states_topic)"/>
    </node>

    <node name="dynamixel_general_control" pkg="dynamixel_general_hw" type="dynamixel_general_control_node" required="$(arg required)" output="screen">
      <remap from="robot_description" to="$(arg robot_description_param)" />
      <remap from="joint_states" to="$(arg joint_states_topic)" />
      <rosparam subst_value="true">
        port_name: $(arg port_name)
        baud_rate: $(arg baud_rate)
        control_rate: $(arg control_rate)
        calculate_effort: $(arg calculate_effort)
      </rosparam>
      <rosparam param="dynamixel_info" file="$(arg dynamixel_info_file)" command="load" />
    </node>

    <rosparam file="$(arg controllers_file)" command="load" />
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"
          args="$(arg controllers_to_start)" />

    <node if="$(arg launch_rviz)"
          name="rviz" pkg="rviz" type="rviz"
          args="-d $(arg rvizconfig)" />

  </group>

</launch>
